### PR TITLE
DAG: Handle lowering unordered compare with inf

### PR DESCRIPTION
**Pull request by arsenm**
Try to take advantage of the nan check behavior of fcmp.
x86_64 looks better, x86_32 looks worse.

### CODE REVIEW

**Comment by arsenm** [comment]
* **#100378** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/100378?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @arsenm and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: llvm/lib/CodeGen/CodeGenCommonISel.cpp [hunk 1]
```diff
@@ -200,6 +201,14 @@ FPClassTest llvm::invertFPClassTestIfSimpler(FPClassTest Test) {
...
   case fcSubnormal | fcZero | fcNan:
     return InvertedTest;
+  case fcInf | fcNan:
+    // If we're trying to use fcmp, we can take advantage of the nan check
+    // behavior of the compare (but this is more instructions in the integer
+    // expansion).
+    return UseFP ? InvertedTest : fcNone;
+  case fcFinite | fcNan:
+    // Inversion of fcInf, which can be done in a combined check.
+    return fcNone;
```
Comment (spavloff) [inline llvm/lib/CodeGen/CodeGenCommonISel.cpp:211]: Was `fcInf` intended here?

File: llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp [hunk 2]
```diff
@@ -8610,26 +8612,36 @@ SDValue TargetLowering::expandIS_FPCLASS(EVT ResultVT, SDValue Op,
...
 
-    if (Test == fcInf &&
-        isCondCodeLegalOrCustom(IsInverted ? ISD::SETUNE : ISD::SETOEQ,
+    bool IsOrderedInf = FPTestMask == fcInf;
+    if ((FPTestMask == fcInf || FPTestMask == (fcInf | fcNan)) &&
+        isCondCodeLegalOrCustom(IsOrderedInf ? OrderedCmpOpcode
+                                             : UnorderedCmpOpcode,
                                 OperandVT.getScalarType().getSimpleVT()) &&
         isOperationLegalOrCustom(ISD::FABS, OperandVT.getScalarType())) {
       // isinf(x) --> fabs(x) == inf
```
Comment (spavloff) [inline llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:8627]: The case `fcInf | fcNan` also deserves comment.

File: llvm/test/CodeGen/X86/is_fpclass.ll [hunk 3]
```diff
@@ -240,18 +240,22 @@ entry:
 define i1 @isfinite_f(float %x) {
 ; X86-LABEL: isfinite_f:
 ; X86:       # %bb.0: # %entry
-; X86-NEXT:    movl $2147483647, %eax # imm = 0x7FFFFFFF
-; X86-NEXT:    andl {{[0-9]+}}(%esp), %eax
-; X86-NEXT:    cmpl $2139095040, %eax # imm = 0x7F800000
-; X86-NEXT:    setl %al
+; X86-NEXT:    flds {{[0-9]+}}(%esp)
```
Comment (spavloff) [inline llvm/test/CodeGen/X86/is_fpclass.ll:243]: It looks like for targets like X86 this behavior is bad. As it requires materialization of infinity value, it produces worse code.

Is there an example when this patch is profitable?

Can we confine this behavior to the profitable cases?

Comment (arsenm) [reply llvm/test/CodeGen/X86/is_fpclass.ll:243]: The x64 case below is fewer instructions. I can't bring myself to care about x86_32 enough to reason about the regressions

File: llvm/include/llvm/CodeGen/CodeGenCommonISel.h [hunk 4]
```diff
@@ -219,9 +219,13 @@ findSplitPointForStackProtector(MachineBasicBlock *BB,
 /// (i.e. fewer instructions should be required to lower it).  An example is the
 /// test "inf|normal|subnormal|zero", which is an inversion of "nan".
 /// \param Test The test as specified in 'is_fpclass' intrinsic invocation.
+///
+/// \param UseFP The intention is to perform the comparison using floating-point
+/// compare instructions which check for nan.
+///
```
[the remark this answers is not in the document]
Comment (arsenm) [reply llvm/include/llvm/CodeGen/CodeGenCommonISel.h:225]: I think they usually separate separate param lines?

File: llvm/lib/CodeGen/CodeGenCommonISel.cpp (hunk 1 above)
Comment (arsenm) [reply llvm/lib/CodeGen/CodeGenCommonISel.cpp:211]: Probably, I think this got lost in one of the many rebases

File: llvm/test/CodeGen/PowerPC/is_fpclass.ll [hunk 5]
```diff
@@ -117,10 +117,12 @@ define i1 @isinf_float(float %x) nounwind {
 define i1 @isinf_ppc_fp128(ppc_fp128 %x) nounwind {
 ; CHECK-LABEL: isinf_ppc_fp128:
 ; CHECK:       # %bb.0:
-; CHECK-NEXT:    xststdcdp 0, 1, 48
-; CHECK-NEXT:    li 3, 0
-; CHECK-NEXT:    li 4, 1
-; CHECK-NEXT:    iseleq 3, 4, 3
+; CHECK-NEXT:    addis 3, 2, .LCPI9_0@toc@ha
```
Comment (chenzheng1030) [inline llvm/test/CodeGen/PowerPC/is_fpclass.ll:120]: This becomes worse for sure.

Comment (ecnelises) [reply llvm/test/CodeGen/PowerPC/is_fpclass.ll:120]: Need special treatment for ppcf128: isfpclas is not legal for ppcf128 but legal for double, however, `isfpclass(ppcf128)` is actually lowered to `isfpclass(double)` (higher half).

File: llvm/test/CodeGen/PowerPC/fp-classify.ll [hunk 6]
```diff
@@ -182,7 +181,8 @@ define <4 x i1> @abs_isinfv4f32(<4 x float> %x) {
 ; P9-NEXT:    xvabssp 0, 34
 ; P9-NEXT:    addi 3, 3, .LCPI6_0@toc@l
 ; P9-NEXT:    lxv 1, 0(3)
-; P9-NEXT:    xvcmpeqsp 34, 0, 1
+; P9-NEXT:    xvcmpeqsp 0, 0, 1
```
Comment (ecnelises) [inline llvm/test/CodeGen/PowerPC/fp-classify.ll:184]: `isCondCodeLegalOrCustom` not working well here? PPC has `xvcmpeqsp` but no `xvcmpnesp`.

File: llvm/test/CodeGen/PowerPC/is_fpclass.ll (hunk 5 above)
Comment (arsenm) [reply llvm/test/CodeGen/PowerPC/is_fpclass.ll]: #105540 + #105577 should fix this

File: llvm/include/llvm/CodeGen/CodeGenCommonISel.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (spavloff) [reply llvm/include/llvm/CodeGen/CodeGenCommonISel.h:226]: In the example in https://llvm.org/docs/CodingStandards.html#doxygen-use-in-documentation-comments parameter lines are not separated by blank lines.

It is not a big deal, but the params separated from each other and NOT separated from the description didn't look good.

**Comment by spavloff** [comment]
Use of the comparison with infinity constant generally is not a good solution unless the target has a cheap way to materialize the infinity, but this comparison has already been implemented in the code and this patch enhances the code generation.

LGTM.

**Review by RKSimon — COMMENTED** [review]
@arsenm Are you happy with the legality check workaround to avoid the x87 regression? We could attempt a custom lowering if you'd prefer?

**Comment by arsenm** [comment]
> @arsenm Are you happy with the legality check workaround to avoid the x87 regression? We could attempt a custom lowering if you'd prefer?

I think it's fine

**Review by RKSimon — APPROVED** [review]
LGTM

**Pull request merged**